pkg/vmprovider/providers/vsphere/session/session_vm_update.go

```diff
@@ -783,3 +783,98 @@ func (s *Session) UpdateVirtualMachine(
+	vmClass, err := utils.GetVirtualMachineClass(vmCtx, client)
+	if err != nil {
+		return nil, err
+	}
+
+	resourcePolicy, err := utils.GetVMSetResourcePolicy(vmCtx, client)
+	if err != nil {
+		return nil, err
+	}
+
+	vmMD, err := utils.GetVMMetadata(vmCtx, client)
+	if err != nil {
+		return nil, err
+	}
+
+	updateArgs.VMClass = vmClass
+	updateArgs.VMMetadata = vmMD
+	if resourcePolicy != nil {
+		updateArgs.ResourcePolicy = resourcePolicy
+	} else {
+		// If resourcePolicy is not set, use default MinCPUFreq value.
+		updateArgs.MinCPUFreq = 0
+	}
+
+	// We're always ready - again - at this point since we've fetched the above objects. We really should
+	// not be touching this condition after creation but that is for another day.
+	conditions.MarkTrue(vmCtx.VM, v1alpha1.VirtualMachinePrereqReadyCondition)
+
+	if lib.IsVMClassAsConfigFSSEnabled() {
+		if cs := updateArgs.VMClass.Spec.ConfigSpec; cs != nil {
+			classConfigSpec, err := utils.GetVMClassConfigSpec(cs)
+			if err != nil {
+				fmt.Printf("updateArgs: %+v\n", updateArgs)
+				return nil, err
+			}
+			updateArgs.ClassConfigSpec = classConfigSpec
+		}
+	}
+
+	imageFirmware := ""
+	// Only get VM image when this is the VM first boot.
+	if isVMFirstBoot(vmCtx) {
+		vmImageStatus, _, err := utils.GetVMImageStatusAndContentLibraryUUID(vmCtx, client)
+		if err != nil {
+			return nil, err
+		}
+		imageFirmware = vmImageStatus.Firmware
+
+		// The only use of this is for the global JSON_EXTRA_CONFIG to set the image name.
+		// The global extra config should only be set during first boot.
+		renderTemplateFn := func(name, text string) string {
+			t, err := template.New(name).Parse(text)
+			if err != nil {
+				return text
+			}
+			b := strings.Builder{}
+			if err := t.Execute(&b, vmImageStatus); err != nil {
+				return text
+			}
+			return b.String()
+		}
+		extraConfig := make(map[string]string, len(updateArgs.ExtraConfig))
+		for k, v := range updateArgs.ExtraConfig {
+			extraConfig[k] = renderTemplateFn(k, v)
+		}
+
+		// Enabling the defer-cloud-init extraConfig key for V1Alpha1Compatible images defers cloud-init from running on first boot
+		// and disables networking configurations by cloud-init. Therefore, only set the extraConfig key to enabled
+		// when the vmMetadata is nil or when the transport requested is not CloudInit.
+		if conditions.IsTrueFromConditions(vmImageStatus.Conditions, v1alpha1.VirtualMachineImageV1Alpha1CompatibleCondition) {
+			extraConfig[constants.VMOperatorV1Alpha1ExtraConfigKey] = constants.VMOperatorV1Alpha1ConfigEnabled
+		}
+		updateArgs.ExtraConfig = extraConfig
+	}
+
+	updateArgs.ConfigSpec = virtualmachine.CreateConfigSpec(
+		vmCtx.VM.Name,
+		&updateArgs.VMClass.Spec,
+		updateArgs.MinCPUFreq,
+		imageFirmware,
+		updateArgs.ClassConfigSpec)
+
+	return updateArgs, nil
+}
+
+func isVMFirstBoot(vmCtx context.VirtualMachineContext) bool {
+	if _, ok := vmCtx.VM.Annotations[FirstBootDoneAnnotation]; ok {
```

Comment:
A devops user can potentially remove the annotation and that could cause the reconcile to break.  Can you file a JIRA story to have some prefix to denote if an annotation is a "system" one?  That way, we can reject any requests to modify the annotation using webhooks.

Reply:
Once this JIRA story https://jira.eng.vmware.com/browse/VMSVC-1261 is finished, this firstboot annotation can be removed. (We'll set image firmware as an annotation) I modified the story description to include webhook part.

Reply:
Can we - at least with the regards to the firmware - ignore that at this point? The class ConfigSpec has gotten confusing in that's theoretically 2 FFS that's effectively consolidated into 1 but the firmware is something that we can or do set when creating the VM and let it be after than.